### PR TITLE
S(dict) -> Dict

### LINKED ISSUE #13425: N(<string>, 30) and sympify: implement a decimal precision kwarg?
Passing a string to `N` is not explicitly documented, but it does work.  The implementation is `sympify(<input>).evalf(n)`.  Should we change `N` to `sympify(<input>, evaluate=False).evalf(n)`?  This improves accuracy in some very specific cases.

This came up downstream https://savannah.gnu.org/bugs/?51748

- - - - -

Here's the background.  The current code generally DTRT, for example:
````
>>> N('sqrt(2)', 30)
1.41421356237309504880168872421
>>> _**2
2.00000000000000000000000000000
````

In general if you put floats in the string, its not going to possible to get more than 15 digits of accuracy and you shouldn't be doing that.  But people do it anyway and then think its a bug:
````
>>> N('2**0.5', 30)
1.41421356237309514547462185874
>>> _**2
2.00000000000000027343234630648
````
But it is possible to get full precision in this case:
````
>>> sympify('2**0.5', evaluate=False).evalf(30)
1.41421356237309504880168872421
>>> _**2
2.00000000000000000000000000000
````

1.  In the current implementation `'2**0.5'` is evaluated to a Float with default precision.  Of course, this cannot be magically re-evaluated in higher precision.
2.  Deferring the evaluation works better in this specific case b/c the float `0.5` is exact in binary.
3.  Deferring will not help at all with `'2**0.1'` because `0.1` is not an integer power of 2 so will be inexact in binary.
4.  I worry this might make some other calculation worse (due to additional rounding error), but I have no concrete example.

### LINKED ISSUE #16769: sympify does not convert dict to Dict
#### Problem
sympify does not convert to dict to Dict which causes hashing problems.

#### Example of problem
```
>>> from sympy import *
>>> d = {1:1}
>>> type(d)
<class 'dict'>
>>> type(sympify(d))
<class 'dict'>
>>> type(Dict(d))
<class 'sympy.core.containers.Dict'>

```
As seen in the above example, dict hasn't been converted to Dict.

#### Other comments/references
Please see [this line's](https://github.com/sympy/sympy/pull/16765/files#diff-d7e4f5f281e7cc9d7e638ad224e148b3R31) comment for issues regarding not converting dict to Dict. 
Reasons for not doing so are awaited.
Thanks in advance.

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

closes #16769
fixes #13425

#### Brief description of what is fixed or changed

- dict can now be sympified
- evaluation of *strings* converts numbers in the string to rational before evaluation (#13425)

#### Release Notes

- core
    - `sympify` will now convert all dict-types input to Dict
    - `N` now uses `rational=True` which gives the desired precision when evaluating string expressions, treating floats as exact

### CODE REVIEW

**Comment by smichr** [comment]
ping @czgdp1807

**Comment by asmeurer** [comment]
> defaultdict is not handled

Does the converter not use `isinstance`/`issubclass`? That should handle defaultdict and OrderedDict, since they are `dict` subclasses.

**Comment by asmeurer** [comment]
Why not the other suggestion at https://github.com/sympy/sympy/issues/13425 (make `N` use `parse_expr`)?

File: sympy/ntheory/factor_.py [hunk 1]
```diff
@@ -1001,7 +1004,7 @@ def factorint(n, limit=None, use_trial=True, use_rho=True, use_pm1=True,
     elif isinstance(n, Mul):
         factordict = {int(k): int(v) for k, v in
             n.as_powers_dict().items()}
-    elif isinstance(n, dict):
+    elif type(n) is dict:
```
Comment (czgdp1807) [inline sympy/ntheory/factor_.py]: Why `type` and not `isinstance`?

File: sympy/core/containers.py [hunk 2]
```diff
@@ -269,6 +269,14 @@ def _sorted_args(self):
         return tuple(sorted(self.args, key=default_sort_key))
 
 
+converter[dict] = lambda d: Dict(*d.items())
```
Comment (czgdp1807) [inline sympy/core/containers.py]: How `converter` works? Is it something inside `SymPy`, like some class which have overrides `__setitem__` and `__getitem__`? 
PS - I am asking this out of curiosity. Thanks in advance.

**Comment by smichr** [comment]
> Does the converter not use `isinstance`/`issubclass`? That should handle defaultdict and OrderedDict, since they are `dict` subclasses.

It does, but `converter[defaultdict]` is set to None so it generates an error:
```python
# a user can pass defaultdict to Dict but the
# result is dict-like not defaultdict-like, so
# by setting the converter to None we will
# catch this case and sympify will raise an error
converter[defaultdict] = None
```

There is no SymPy equivalent of a defaultdict. I can handle it like a list, however, and send back a defaultdict with everything sympified and with the factory producing sympified values.

File: sympy/core/containers.py (hunk 2 above)
Comment (smichr) [reply sympy/core/containers.py]: It's a dicstionary that can be imported where it is need (like in `sympify` that has a key that is a cls/type and value that is a unary function that converts the passed object into a SymPy object.

File: sympy/ntheory/factor_.py (hunk 1 above)
Comment (smichr) [reply sympy/ntheory/factor_.py]: Unintentional change.

**Comment by smichr** [comment]
> Why not the other suggestion at #13425 (make `N` use `parse_expr`)?

Because then the following duplication of code is not needed:

```
    from sympy.parsing.sympy_parser import (parse_expr, TokenError,
                                            standard_transformations as T)
    if isinstance(expr, string_types):
      try:
            x = x.replace('\n', '')
            expr = parse_expr(x, transformations=T, evaluate=False)
      except (TokenError, SyntaxError) as exc:
            raise SympifyError('could not parse %r' % x, exc)
    else:
        expr = x
    return expr.evalf(n, **options)
```

But if you prefer, I can put this there.

File: sympy/ntheory/factor_.py (hunk 1 above)
Comment (smichr) [inline sympy/ntheory/factor_.py]:
```suggestion
    elif isinstance(n, dict):
```

File: sympy/core/containers.py [hunk 3]
```diff
@@ -269,6 +269,14 @@ def _sorted_args(self):
...
+# a user can pass defaultdict to Dict but the
+# result is dict-like not defaultdict-like, so
+# by setting the converter to None we will
+# catch this case and sympify will raise an error
+converter[defaultdict] = None
+
```
Comment (smichr) [inline sympy/core/containers.py]:
```suggestion
converter[OrderedDict] = None  # the same applies for OrderedDict
```

**Comment by asmeurer** [comment]
I don't remember the details from #13425. I guess there were some concerns about evaluate=False breaking N in other ways because of other expressions not being evaluated. I don't think it's a big deal.

File: sympy/core/containers.py [hunk 4]
```diff
@@ -269,6 +269,14 @@ def _sorted_args(self):
         return tuple(sorted(self.args, key=default_sort_key))
 
 
+converter[dict] = lambda d: Dict(*d.items())
+# a user can pass defaultdict to Dict but the
+# result is dict-like not defaultdict-like, so
+# by setting the converter to None we will
+# catch this case and sympify will raise an error
+converter[defaultdict] = None
+converter[OrderedDict] = None  # the same applies for OrderedDict
```
Comment (asmeurer) [inline sympy/core/containers.py]: I guess I can understand disabling defaultdict (though honestly I would just let it work), but OrderedDict is the same as dict, except ordered. And sympy.Dict is immutable and ordered, so I think it's fine.

File: sympy/core/evalf.py [hunk 5]
```diff
@@ -1536,4 +1536,4 @@ def N(x, n=15, **options):
     1.291
 
     """
-    return sympify(x).evalf(n, **options)
+    return sympify(x, evaluate=False).evalf(n, **options)
```
Comment (smichr) [inline sympy/core/evalf.py]:
```suggestion
    return sympify(x, rational=True).evalf(n, **options)  # rational will help string case
```

**Comment by smichr** [comment]
> concerns about evaluate=False breaking

The rational=True flag (rather than evaluate=False) is used which fixes the other issues that were raised at #13425.

File: sympy/core/containers.py (hunk 4 above)
Comment (smichr) [reply sympy/core/containers.py]: > I would just let it work

all dict-types are handled the same way, now

**Pull request merged**